# Model Evaluation


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
# pd.options.display.float_format = '{:,.2f}'.format
%matplotlib inline
from pycaret.regression import *
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.metrics import mean_squared_error, r2_score


In [12]:
Crop_df=pd.read_csv("data\Crop_data_filtered.csv")

In [13]:
Crop_df

State_Name District_Name  Crop_Year      Season  \
0       Andaman_and_Nicobar_Islands      NICOBARS       2000      Kharif   
1       Andaman_and_Nicobar_Islands      NICOBARS       2000      Kharif   
2       Andaman_and_Nicobar_Islands      NICOBARS       2000      Kharif   
3       Andaman_and_Nicobar_Islands      NICOBARS       2000  Whole_Year   
4       Andaman_and_Nicobar_Islands      NICOBARS       2000  Whole_Year   
...                             ...           ...        ...         ...   
229373                  West_Bengal       PURULIA       2014      Summer   
229374                  West_Bengal       PURULIA       2014      Summer   
229375                  West_Bengal       PURULIA       2014  Whole_Year   
229376                  West_Bengal       PURULIA       2014      Winter   
229377                  West_Bengal       PURULIA       2014      Winter   

                       Crop      Area  Production        Zone Crop_Category  \
0                  Arecanut    1254.0      2000.0    ut_india        Others   
1       Other_Kharif_pulses       2.0         1.0    ut_india        Others   
2                      Rice     102.0       321.0    ut_india       Cereals   
3                    Banana     176.0       641.0    ut_india        Fruits   
4                 Cashewnut     720.0       165.0    ut_india          Nuts   
...                     ...       ...         ...         ...           ...   
229373                 Rice     306.0       801.0  east_india       Cereals   
229374              Sesamum     627.0       463.0  east_india      Oilseeds   
229375            Sugarcane     324.0     16250.0  east_india     Sugarcane   
229376                 Rice  279151.0    597899.0  east_india       Cereals   
229377              Sesamum     175.0        88.0  east_india      Oilseeds   

        Production/Area  
0              1.594896  
1              0.500000  
2              3.147059  
3              3.642045  
4              0.229167  
...                 ...  
229373         2.617647  
229374         0.738437  
229375        50.154321  
229376         2.141848  
229377         0.502857  

[229378 rows x 10 columns]

In [14]:
Crop_df.Crop.unique()

array(['Arecanut', 'Other_Kharif_pulses', 'Rice', 'Banana', 'Cashewnut',
       'Coconut', 'Dry_ginger', 'Sugarcane', 'Sweet_potato', 'Tapioca',
       'Black_pepper', 'Dry_chillies', 'other_oilseeds', 'Turmeric',
       'Maize', 'MoongGreen_Gram', 'Urad', 'ArharTur', 'Groundnut',
       'Sunflower', 'Bajra', 'Castor_seed', 'Cottonlint', 'Horse-gram',
       'Jowar', 'Korra', 'Ragi', 'Gram', 'Wheat', 'Tobacco', 'Masoor',
       'Sesamum', 'Linseed', 'Safflower', 'Onion', 'other_misc._pulses',
       'Samai', 'Small_millets', 'Coriander', 'Potato',
       'Other__Rabi_pulses', 'Beans_&_MutterVegetable', 'Bhindi',
       'Brinjal', 'Citrus_Fruit', 'Grapes', 'Mango', 'Orange',
       'Other_Fresh_Fruits', 'Papaya', 'Pome_Fruit', 'Tomato', 'Soyabean',
       'Mesta', 'CowpeaLobia', 'Lemon', 'Pome_Granet', 'Sapota',
       'Cabbage', 'Rapeseed_&Mustard', 'Niger_seed', 'Varagu', 'Garlic',
       'Ginger', 'Oilseeds_total', 'Pulses_total', 'Jute',
       'Peas_&_beans_Pulses', 'Blackgram', 'P

In [15]:
Crop_df=Crop_df.drop(['District_Name','Zone','Crop_Category'],axis=1)

In [16]:
le=LabelEncoder()
categorical_col=Crop_df.select_dtypes(include="object")
for i in categorical_col:
    Crop_df[i]=le.fit_transform(Crop_df[i])

In [17]:
X=Crop_df.drop("Production",axis=1)
y=Crop_df["Production"]

In [18]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.33,random_state=1)


In [19]:
scaler=StandardScaler()
x_train_scaled=scaler.fit_transform(X_train)
x_test_scaled=scaler.transform(X_test)


In [20]:
# Step 3: Create and train the ExtraTreeRegressor model
model = ExtraTreesRegressor(n_estimators=100, random_state=42)
model.fit(x_train_scaled, y_train)

# Step 4: Make predictions on the test set
y_pred = model.predict(x_test_scaled)

# Step 5: Evaluate the model
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print("Root Mean Squared Error (RMSE):", rmse)
print("R-squared (R2) Score:", r2)


Root Mean Squared Error (RMSE): 677747.9400213145
R-squared (R2) Score: 0.9984469846533012


In [21]:
df_comparison = pd.DataFrame({'Original': y_test, 'Predicted': y_pred})
print(df_comparison.sample(20))

         Original     Predicted
217510        9.0  9.020000e+00
95307     23084.0  2.314212e+04
85092       161.0  1.617260e+02
41289     40092.0  3.995401e+04
37287       120.0  1.185400e+02
185534       64.0  6.404000e+01
32590    108852.0  1.090195e+05
54219       100.0  1.004990e+02
40385      2079.0  2.096700e+03
72244     15845.0  1.595175e+04
167310     2715.0  2.725613e+03
189582     7212.0  7.288858e+03
125154     2400.0  2.404664e+03
114085    72100.0  7.213227e+04
125085     1585.0  1.580797e+03
33560       845.0  8.399800e+02
45596       267.0  2.694680e+02
1348       3000.0  3.000000e+03
212095  2567192.0  2.603889e+06
131627     6186.0  6.170480e+03


### In this Jupyter Notebook project, we utilized the ExtraTreeRegressor model for our predictive analysis. The model's performance was evaluated using key metrics such as Root Mean Squared Error (RMSE) and R-squared (R2) score, which were found to be 677,747.94 and 0.9984 respectively. These metrics provide insights into the accuracy and predictive power of the ExtraTreeRegressor model in our project.

**Project Summary and Conclusions**

**1. Introduction:**
   The analysis of agricultural data aimed to uncover production trends, crop distribution patterns, and seasonal variations across different regions and states. The focus was primarily on understanding the dynamics of crop production to aid in decision-making processes for stakeholders in the agricultural sector.

**2. Insights from Tableau Dashboard:**

   - **Production Distribution:** 
     - Uttar Pradesh, Maharashtra, and Tamil Nadu collectively contribute nearly 50% of the total production.
     - Sugarcane dominates crop production, contributing to 48.80% of the total, followed by rice and wheat.
     
   - **Seasonal Analysis:** 
     - Kharif season accounts for the highest share of production (37.79%), followed by the entire year (37.41%), Rubi (18.66%), winter (3.98%), and summer (1.58%).
     
   - **State-specific Insights:** 
     - Uttar Pradesh's major crops include sugarcane, wheat, rice, and potato, with significant production during the kharif season.
     - Maharashtra's production is dominated by sugarcane, cotton lint, soybean, and rice, with considerable production throughout the year.
     
   - **Crop-wise Analysis:** 
     - Major crops during the kharif season include sugarcane, rice, cotton lint, maize, and jute, while sugarcane, potato, and banana dominate throughout the year.
     
   - **Crop Yield and Yearly Trends:** 
     - Sugarcane, papaya, and banana have the highest yield per hectare. Top production years vary across different crops and regions.
     
   - **Dataset Completeness:** 
     - Data availability varies across different years, with 2002 having the highest availability of crop data.
     
   - **Regional Production:** 
     - In 2014, kharif season production accounted for a significant portion of the total production, with North India contributing the highest share.


**3. Analysis in Jupyter Notebook:**

   - **Data Preprocessing:** 
     - Data cleaning involved dropping unnecessary columns and encoding categorical variables for model training.
     
   - **Model Training and Evaluation:** 
     - Utilized ExtraTreesRegressor model for predicting production values.
     - Achieved a high R-squared score of 0.9984, indicating excellent model fit. However, the relatively high RMSE suggests room for improvement in predictive accuracy.


     
**4. Conclusions:**
   - The analysis revealed significant insights into production trends, crop distribution, and seasonal variations, providing valuable information for stakeholders in the agricultural sector.
   - While the model demonstrated a high degree of explanatory power, further refinement may be necessary to improve predictive accuracy and address potential outliers or data variability.
   - Overall, the project contributes to a better understanding of agricultural dynamics and supports informed decision-making for agricultural planning and management.

## Thank you 
Akash Patil<br>
aakashgolu1008@gmail.com